### PR TITLE
Add OpInfo test and fix a couple cases

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68629
* #68628
* #67045
* #67041
* #67820
* __->__ #66294
* #66293
* #66292

In this PR:
- OpInfo for forward AD now checks batched forward grad when `op.check_batched_grad=True`
- Adds setting to disable the test for individual ops `check_batched_forward_grad` and disable for the ops here: https://github.com/pytorch/pytorch/issues/66357

Fixes some more failures:
- Make Forward AD metadata less strict by allowing stride to differ when size is 1
- Fix sum batching rule when logical tensor is a scalar and dim is unspecified
- Batching rule for `_reshape_alias`
- ~Batching rules now preserve storage offset for view operator that return non-zero storage offset~ (moved to previous PR)

Differential Revision: [D31842020](https://our.internmc.facebook.com/intern/diff/D31842020)

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
I'm confused by the PR title.
Does this fix any test?

**Review by albanD — COMMENTED** [review]
Looks mostly good.
cc @zou3519 in case you want to double check the vmap changes.

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842020).

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -506,6 +506,7 @@ def __init__(self,
                                            # defaults to support_autograd's value
                  check_batched_gradgrad=None,  # whether to check batched grad grad when doing gradgradcheck
                                                # default's to support_gradgrad's value
+                 check_batched_forward_grad=None,
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:509]: Should this one have a comment like the others?

**Review by albanD — APPROVED** [review]
Doc update but lgtm otherwise

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842020).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842020).

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 2]
```diff
@@ -63,7 +63,7 @@ Tensor sum_batching_rule(const Tensor& self, IntArrayRef dims, bool keepdim, opt
   // >>> x = torch.randn(B0)  # the per-examples are all scalars
   // >>> vmap(partial(torch.sum, dim=0), x)
   // then we replicate the behavior of sum(scalar_tensor, dim=0).
-  if (/*logical*/self.dim() == 0 && dims.size() == 1 && is_allowed_dim_on_scalar_tensor(dims[0])) {
+  if (/*logical*/self.dim() == 0 && (dims.size() == 0 || (dims.size() == 1 && is_allowed_dim_on_scalar_tensor(dims[0])))) {
```
Comment (zou3519) [inline aten/src/ATen/BatchingRegistrations.cpp:66]: Add a test in test_vmap maybe?

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 3]
```diff
@@ -542,6 +542,14 @@ static void checkBasicAsStridedValidForSlice(
       "rewrite the `as_strided` call as a sequence of PyTorch view operations");
 }
 
+Tensor _reshape_alias_batching_rule(const Tensor& self, IntArrayRef sizes, IntArrayRef strides) {
+  auto self_physical = MultiBatchVmapTransform::logicalToPhysical(self);
+  auto physical_shape = self_physical.getPhysicalShape(sizes);
+  auto physical_strides = self_physical.getPhysicalStrides(strides);
+  auto result = self_physical.tensor()._reshape_alias(physical_shape, physical_strides);
+  return self_physical.getPhysicalToLogicalMap().apply(result);
+}
```
Comment (zou3519) [inline aten/src/ATen/BatchingRegistrations.cpp:551]: Could we do the following instead?
```
Tensor _reshape_alias_batching_rule(const Tensor& self, IntArrayRef sizes, IntArrayRef strides) {
  return reshape_batch_rule(self, sizes);
}
```

File: aten/src/ATen/VmapTransforms.cpp [hunk 4]
```diff
@@ -80,6 +80,17 @@ VmapDimVector VmapPhysicalView::getPhysicalShape(IntArrayRef logical_shape) cons
...
+  at::VmapDimVector physical_strides;
+  auto batch_strides = tensor().strides().slice(0, numBatchDims());
+  physical_strides.reserve(numBatchDims() + logical_strides.size());
+  physical_strides.insert(
+      physical_strides.end(), batch_strides.begin(), batch_strides.end());
+  physical_strides.insert(
+      physical_strides.end(), logical_strides.begin(), logical_strides.end());
+  return physical_strides;
+}
+
```
Comment (zou3519) [inline aten/src/ATen/VmapTransforms.cpp:93]: This should not be a helper function. It's possible that the strides of the batch dims don't "match" the logical strides. Like if you have a tensor of size(2, 3, 4) and all dimensions are batch dims, and one passes in logical_strides(5, 1), then we're going to end up creating a tensor with strides (12, 4, 1, 5, 1) which is bogus: what the user might have wanted is a tensor with stride (2 * 3 * 4 * 5, 3 * 4 * 5, 4 * 5, 5, 1)

**Review by zou3519 — COMMENTED** [review]
We should not have a getPhysicalStrides because that doesn't make much sense. I suggested a rewrite for the `_reshape_alias` batch rule to avoid getPhysicalStrides.

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:551]: Or call reshape instead of reshape_alias here

**Review by zou3519 — APPROVED** [review]

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842020).

**Review by albanD — COMMENTED** [review]
CI is not good.
Also given the CI logs, we need to silence the vmap warning in the test_ops as it is (way too) spammy.

Looks good otherwise!

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842020).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842020).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31842020).
